Clean Read Only Geth

hrthaowang opened the pull request:
In this PR, I removed the `readOnlyNode` and `readOnlyBackend`. Instead, I used a boolean variable to represent whether it's in the `readOnly` mode to make the code cleaner.

holiman committed:
core/rawdb: do prefixed lookup first

karalabe committed:
Merge pull request #24288 from holiman/prefer_prefixed

core/rawdb: do prefixed lookup first

MariusVanDerWijden committed:
core/rawdb: fix typo (#24289)

holiman committed:
eth/tracers: avoid using blockCtx concurrently (#24286)

hrthaowang committed:
test

hrthaowang committed:
test

hrthaowang committed:
clean code

hrthaowang committed:
clean code

s1na committed:
cmd/utils: open db in readonly when dev datadir exists (#24298)

s1na committed:
internal/web3ext: add eth.getLogs wrapper (#24297)

ValentinTrinque committed:
rpc: set Request.GetBody for client requests (#24292)

When talking to an HTTP2 server, there are situations where it needs to
"rewind" the Request.Body. To allow this, we have to set up the Request.GetBody
function to return a brand new instance of the body.

If not set, we can end up with the following error:

    http2: Transport: cannot retry err [http2: Transport received Server's graceful shutdown GOAWAY] after Request.Body was written; define Request.GetBody to avoid this error

See this commit for more information: https://sourcegraph.com/github.com/golang/net/-/commit/cffdcf672aee934982473246bc7e9a8ba446aa9b?visible=2

danielmh0 commented on core/rawdb/accessors_state.go:
```diff
@@ -41,16 +41,14 @@ func WritePreimages(db ethdb.KeyValueWriter, preimages map[common.Hash][]byte) {
-	// Try with the legacy code scheme first, if not then try with current
-	// scheme. Since most of the code will be found with legacy scheme.
-	//
-	// todo(rjl493456442) change the order when we forcibly upgrade the code
```
Why did you need this change, and are we sure it's safe?

danielmh0 commented on eth/backend.go:
```diff
@@ -131,12 +133,28 @@ func New(stack *node.Node, config *ethconfig.Config) (*Ethereum, error) {
+		genesisHash = common.Hash{}
+		genesisErr = nil
+	} else {
+		chainDb, err = stack.OpenDatabaseWithFreezer("chaindata", config.DatabaseCache, config.DatabaseHandles, config.DatabaseFreezer, "eth/db/chaindata/", false)
```
If the database cannot be opened read-only with a freezer, why does `OpenDatabaesWithFreezer` have a `readonly` arg? Please comment this for the reader.

danielmh0 commented on eth/backend.go:
```diff
@@ -196,12 +215,18 @@ func New(stack *node.Node, config *ethconfig.Config) (*Ethereum, error) {
 			Preimages:           config.Preimages,
 		}
 	)
+	if readOnly {
+		cacheConfig.SnapshotLimit = 0
+		cacheConfig.TrieDirtyDisabled = true
+		cacheConfig.TrieCleanJournal = ""
+		config.TxPool.Journal = ""
+	}
```
These params should have been set appropriately in `makeConfigNode` from `readOnlyGeth/config.go`, rather than changed after-the-fact here.

danielmh0 commented on eth/backend.go:
```diff
@@ -538,46 +568,51 @@ func (s *Ethereum) Protocols() []p2p.Protocol {
-	if s.config.LightServ > 0 {
-		if s.config.LightPeers >= s.p2pServer.MaxPeers {
-			return fmt.Errorf("invalid peer config: light peer count (%d) >= total peer count (%d)", s.config.LightPeers, s.p2pServer.MaxPeers)
+	if s.readOnly {
```
Shouldn't this be `!s.readOnly`?

danielmh0 commented on eth/backend.go:
```diff
@@ -538,46 +568,51 @@ func (s *Ethereum) Protocols() []p2p.Protocol {
-	s.handler.Stop()
-
+	// Only Stop P2P when P2P was initiated
+	if !s.readOnly {
+		// Stop all the peer-related stuff first.
+		s.ethDialCandidates.Close()
+		s.snapDialCandidates.Close()
+		s.handler.Stop()
+	}
```
I don't know if this is easy to code, but it'd be nice not to branch on `!s.readOnly` in particular, but rather have the `Close()`/`Stop()` functions handle a never-opened/never-started case more gracefully.

Analogous to how the destructor of a unique_ptr just does the right thing even if the unique_ptr is null.

danielmh0 commented on node/node.go:
```diff
@@ -104,52 +108,56 @@ func New(conf *Config) (*Node, error) {
-	node.keyDir = keyDir
-	node.keyDirTemp = isEphem
+
 	// Creates an empty AccountManager with no backends. Callers (e.g. cmd/geth)
 	// are required to add the backends later on.
 	node.accman = accounts.NewManager(&accounts.Config{InsecureUnlockAllowed: conf.InsecureUnlockAllowed})
```
Can we try to keep all the `!node.readOnly` logic contained in fewer `if` blocks? It seems like we could merge above and below, if this assignment was moved up?

danielmh0 commented on node/node.go:
```diff
@@ -342,50 +352,52 @@ func (n *Node) startRPC() error {
 	if err := n.startInProc(); err != nil {
```
I think it's cleaner to easier to rename this function `startExternalRPC`, remove `n.startInProc()` from it entirely, and avoid extra layer of branching on `!n.readOnly`.

Overall: You have a lot of `if !n.readOnly` branches, which is a bad code smell -- try to refactor the code so that the read-only special case is self-contained.

danielmh0 commented on node/node.go:
```diff
@@ -396,9 +408,11 @@ func (n *Node) wsServerForPort(port int) *httpServer {
 }
 
 func (n *Node) stopRPC() {
```
Similar to above, I think this should be `stopExternalRPC`.

danielmh0 commented on node/node.go:
```diff
@@ -396,9 +408,11 @@ func (n *Node) wsServerForPort(port int) *httpServer {
-	n.http.stop()
-	n.ws.stop()
-	n.ipc.stop()
+	if !n.readOnly {
```
Here and elsewhere, you're conflating "read-only" with "used-as-a-library" in a few places. You could imagine starting HTTP/WS/IPC services on a read-only geth process. Maybe `localLibrary` should be a separate option in Node?

And `readOnlyGeth/` should be renamed `readOnlyLib/`?

chfast committed:
tests: external evm benchmarks (#24050)

* tests: add ipsilon/evm-benchmarks git submodule

* tests: plug-in evm-benchmarks

ligi committed:
cmd/geth: make test pass on a pi4 by using lightkdf (#24314)

s1na committed:
eth/tracers: native prestate tracer (#24268)

* eth/tracers: add initial native prestate tracer

* fix balance hex

* handle prestate for tx from and to

* drop created contract from prestate

* fix sender balance

* use switch instead

Co-authored-by: Martin Holst Swende <martin@swende.se>

* minor fix

* lookup create2 account

* mv code around a bit

* check stackLen for create2

* fix transfer tx for js prestate tracer

* fix create2 addr

* track extcodehash in js prestate tracer

Co-authored-by: Martin Holst Swende <martin@swende.se>

ligi committed:
build: fix lint on ARM (#24311)

* build: append GOARM to arm lint download URL

otherwise it fails with:

downloading from https://github.com/golangci/golangci-lint/releases/download/v1.42.0/golangci-lint-1.42.0-linux-arm.tar.gz
ci.go:347: download error: status 404

* build: increase timeout for lint

Otherwise it times out on a pi

* Increase timeout even further

saw longer build times

rjl493456442 committed:
all: separate catalyst package (#24280)

* all: seperate catalyst package

* eth/catalyst: moved some methods, added docs

* eth/catalyst, les/catalyst: add method docs

* core, eth, les, miner: move common function to beacon package

* eth/catalyst: goimported

* cmd/utils, miner/stress/beacon: naming nitpicks

Co-authored-by: Marius van der Wijden <m.vanderwijden@live.de>
Co-authored-by: Péter Szilágyi <peterke@gmail.com>

B-Harden committed:
accounts: correct spelling mistake (#24323)

I believe the sentence is attempting to explain that the URL is "[used] by upper layers to define a sorting order over all wallets from multiple backends."

s1na committed:
eth/tracers: clean-up tracer collection (#24320)

* eth/tracers: clean-up tracer collection

* Rm test for dropped tracer

komkom committed:
accounts/abi/bind: improve WaitMined error handling (#24321)

This change makes it so WaitMined no longer logs an error when the receipt
is unavailable. It also changes the simulated backend to return NotFound for
unavailable receipts, just like ethclient does.

karalabe committed:
cmd: auto-enable beacon APIs if TTD is defined

karalabe committed:
Merge pull request #24328 from karalabe/uke-catalyst

cmd: auto-enable beacon APIs if TTD is defined

aaronbuchwald committed:
go.mode: bump graphql-go dependency to v1.3.0 (#24324)

holiman committed:
cmd/devp2p: implement snap protocol testing (#24276)

This also contains some changes to the protocol handler to
make the tests pass.

haowang0402 committed:
add LocalLib option

rjl493456442 committed:
miner: avoid data race in miner (#24349)

haowang0402 committed:
fix style

haowang0402 committed:
fix style

haowang0402 committed:
expose ServeRawRequset Under rpc.Server

haowang0402 committed:
expose ServeRawRequset Under rpc.Server

hrthaowang committed:
add back debug logger

hrthaowang committed:
test stuff

hrthaowang committed:
Merge branch 'clean-read-only' of https://github.com/hudson-trading/go-ethereum into clean-read-only

hrthaowang committed:
fix merge bug
